File: 0680-valid-palindrome-ii/0680-valid-palindrome-ii.java
class Solution {
    public boolean validPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;
        while(i < j){
            if(s.charAt(i) != s.charAt(j)){
                return check(s , i+1 , j) || check(s , i , j-1);
            }
            else {
                i++;
                j--;
            }
        }
        return true;
    }
    public boolean check(String s , int i , int j){
        while(i < j){
            if(s.charAt(i) != s.charAt(j)) return false;
            i++;j--;
        }
        return true;
    }
}

File: 0680-valid-palindrome-ii/0680-valid-palindrome-ii.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s)- 1
        while left < right:
            if s[left] != s[right]:
                return self.checkPalindromeItr(s, left+1, right) or self.checkPalindromeItr(s, left, right-1)
            left += 1
            right -= 1
        return True
        
                
    def checkPalindromeItr(self, s:str, left:int, right:int) -> bool:
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True
    
    def checkPalindrome(self, s:str, left:int, right: int) -> int:
        if left >= right:
            return 0
        c = s[left]
        d = s[right]
        if c == d:
            return self.checkPalindrome(s, left+1, right-1)
        return 1 + min(self.checkPalindrome(s, left+1, right), self.checkPalindrome(s, left, right-1))
        
        